### ISSUE TITLE
Add code fix provider to convert assertions to new API

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#1297]
#1292 added a new assertion API, superseding the old version that had methods using the `Repeated` class.
The analyzer could help people convert the old uses into new ones.

One danger of doing this is that someone may use a FakeItEasy analyzer from the future, that is, a version with the code fix provider, while targeting a FakeItEasy version that doesn't contain the new API.
It's probably not very likely, so we should consider a way to reduce the risk. Perhaps (and I'm just thinking out loud here) we could interrogate the `IAssertConfiguration` instance to see if it has the new methods before offering to convert to them.

**Comment by thomaslevesque** [comment]
> Perhaps (and I'm just thinking out loud here) we could interrogate the `IAssertConfiguration` instance to see if it has the new methods before offering to convert to them.

Yes, I think we can do this.

**Comment by thomaslevesque** [comment]
> blairconrad added the in-progress label 11 hours ago

Wow, already?

**Comment by blairconrad** [comment]
> Wow, already?

Oh, :blush:. I figured I could always un-progress it if someone disagreed.

**Comment by blairconrad** [comment]
I'd like to offer the "nicest" options possible for code fixes, that is, I'd like to offer these fixes, where the user could choose the format they wanted when there are multiple conversions from a given source assertion:

```c#
MustHaveHappened(Repeated.Never) → MustNotHaveHappened()

MustHaveHappened(Repeated.Exactly.Once)    → MustHaveHappenedOnceExactly()
MustHaveHappened(Repeated.AtLeast.Once)    → MustHaveHappenedOnceOrMore()
MustHaveHappened(Repeated.NoMoreThan.Once) → MustHaveHappenedOnceOrLess()

MustHaveHappened(Repeated.Exactly.Twice)    → MustHaveHappenedTwiceExactly()
MustHaveHappened(Repeated.AtLeast.Twice)    → MustHaveHappenedTwiceOrMore()
MustHaveHappened(Repeated.NoMoreThan.Twice) → MustHaveHappenedTwiceOrLess()

MustHaveHappened(Repeated.Exactly.Times(n))    → MustHaveHappened(n, Times.Exactly)
MustHaveHappened(Repeated.AtLeast.Times(n))    → MustHaveHappened(n, Times.OrMore)
MustHaveHappened(Repeated.NoMoreThan.Times(n)) → MustHaveHappened(n, Times.OrLess)

MustHaveHappened(Repeated.Exactly.Times(0)) → MustHaveHappened(0, Times.Exactly)
MustHaveHappened(Repeated.Exactly.Times(0)) → MustNotHaveHappened()

MustHaveHappened(Repeated.AtLeast.Times(0)) → MustHaveHappened(0, Times.OrMore)
// although maybe that should be flagged as an error

MustHaveHappened(Repeated.NoMoreThan.Times(0)) → MustNotHaveHappened()
MustHaveHappened(Repeated.NoMoreThan.Times(0)) → MustHaveHappened(0, Times.OrLess)

MustHaveHappened(Repeated.Exactly.Times(1)) → MustHaveHappenedOnceExactly()
MustHaveHappened(Repeated.Exactly.Times(1)) → MustHaveHappened(1, Times.Exactly)

MustHaveHappened(Repeated.AtLeast.Times(1)) → MustHaveHappened()
MustHaveHappened(Repeated.AtLeast.Times(1)) → MustHaveHappenedOnceOrMore()
MustHaveHappened(Repeated.AtLeast.Times(1)) → MustHaveHappened(1, Times.OrMore)

MustHaveHappened(Repeated.NoMoreThan.Times(1)) → MustHaveHappenedOnceOrLess()
MustHaveHappened(Repeated.NoMoreThan.Times(1)) → MustHaveHappened(1, Times.OrLess)

MustHaveHappened(Repeated.Exactly.Times(2)) → MustHaveHappenedTwiceExactly()
MustHaveHappened(Repeated.Exactly.Times(2)) → MustHaveHappened(2, Times.Exactly)

MustHaveHappened(Repeated.AtLeast.Times(2)) → MustHaveHappenedTwiceOrMore()
MustHaveHappened(Repeated.AtLeast.Times(2)) → MustHaveHappened(2, Times.OrMore)

MustHaveHappened(Repeated.NoMoreThan.Times(2)) → MustHaveHappenedTwiceOrLess()
MustHaveHappened(Repeated.NoMoreThan.Times(2)) → MustHaveHappened(2, Times.OrLess)

MustHaveHappened(Repeated.Like(n => n % 2 == 0)  → MustHaveHappenedANumberOfTimesMatching(n => n % 2 == 0)
```

Too ambitious? Too confusing? Alternatively, we could either go for the straightforward code fix, or only ever offer what we think of as the "nicest" fix.

**Comment by thomaslevesque** [comment]
> Too ambitious? Too confusing? Alternatively, we could either go for the straightforward code fix, or only ever offer what we think of as the "nicest" fix.

I would just go for the straightforward fix. If you feel like offering other options, why not, but I don't think it's really necessary.

**Comment by thomaslevesque** [comment]
This change has been released as part of [FakeItEasy 4.5.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/4.5.0).
